### PR TITLE
ENG-9560: AAUIW a higher limit on maximum AdHoc query length

### PR DESCRIPTION
**Pull request by JadenW**
Seems only need to change serialize/deserialize part of AdHocPlannedStatement.
No need to change ParameterSet, not affect wire protocol.

### CODE REVIEW

**Comment by akhanzode** [comment]
If jenkins is happy merge it.

File: src/ee/execution/VoltDBEngine.cpp [hunk 1]
```diff
@@ -507,11 +502,19 @@ void VoltDBEngine::resetExecutionMetadata() {
 
     // set this back to -1 for error handling
     m_currentInputDepId = -1;
+
     if (m_currExecutorVec == NULL) {
         // This is usually the result of some planner error producing an
         // invalid plan that can not be converted into plan nodes / executors.
         return;
     }
+
```
Comment (akhanzode) [inline src/ee/execution/VoltDBEngine.cpp]: Why this change is needed?

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ public synchronized static AdHocCompilerCache getCacheForCatalogHash(byte[] cata
 
     // cache sizes determined at construction time
     final int MAX_LITERAL_ENTRIES;
+    final long MAX_LITERAL_MEM;
```
Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: Can you make this into a hidden tunable via system property. Do you need the number of entries now that its bounded by memory?

File: src/ee/execution/VoltDBEngine.cpp (hunk 1 above)
Comment (JadenW) [inline src/ee/execution/VoltDBEngine.cpp]: m_executorContext could be invalid if planner hit some error. 
So It should first check m_currExecutorVec == NULL before check m_executorContext

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 2 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: I now made this cache bound by both memory and number of entries. 
We still don't want too many cache entries for small size queries, do we? 
Shoud I drop the entries limit?

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 3]
```diff
@@ -147,9 +138,73 @@ public boolean removeEldestEntry(Map.Entry<String, List<BoundPlan> > eldest) {
...
+    // define a LinkedHashMap based LRU cache bounds by both entry number and entry value on-heap size
+    // without changing Map.Entry, only works for value of type AdHocPlannedStatement
+    // only extend put, remove,clear and removeEldestEntry methods to account weight
+    public class MemoryBoundedLinkedHashMap<K, AdHocPlannedStatement> extends LinkedHashMap<K, AdHocPlannedStatement> implements Map<K, AdHocPlannedStatement>{
+       private static final long serialVersionUID = 2988383448026641836L;
+        private final int maxEntries;
+        private final long maxMemory; // in bytes
+        private long currentMemory;   // in bytes
+
+        public long getCurrentMemory() {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: Do you need this method? seems like unused.

File: src/frontend/org/voltdb/ParameterSet.java [hunk 4]
```diff
@@ -255,6 +255,9 @@ public static ParameterSet fromByteBuffer(ByteBuffer buffer) throws IOException
         int startPos = buffer.position();
 
         short count = buffer.getShort();
+        if (count < 0) {
+            throw new RuntimeException("ParameterSet exceeds the length limitation " + Short.MAX_VALUE);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/ParameterSet.java]: The error message is wrong it should be invalid parameter length. Also add count in message.

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 5]
```diff
@@ -147,9 +138,73 @@ public boolean removeEldestEntry(Map.Entry<String, List<BoundPlan> > eldest) {
...
                 return false;
             }
+
         };
     }
 
+    // define a LinkedHashMap based LRU cache bounds by both entry number and entry value on-heap size
+    // without changing Map.Entry, only works for value of type AdHocPlannedStatement
+    // only extend put, remove,clear and removeEldestEntry methods to account weight
+    public class MemoryBoundedLinkedHashMap<K, AdHocPlannedStatement> extends LinkedHashMap<K, AdHocPlannedStatement> implements Map<K, AdHocPlannedStatement>{
```
Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: Can you move this class in its own file so if anyone wants to use it they can. Also you can write unit test for just this memory bounded LHM.

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 6]
```diff
@@ -147,9 +138,73 @@ public boolean removeEldestEntry(Map.Entry<String, List<BoundPlan> > eldest) {
...
+            // default max value size of 32MB
+            this(1000, 32 * 1024 * 1024);
+        }
+
+        public MemoryBoundedLinkedHashMap(final int maxEntries) {
+            this(maxEntries, 32 * 1024 * 1024);
+        }
+
+        public MemoryBoundedLinkedHashMap(final int maxEntries, final long maxMemory) {
+            super(maxEntries * 2, .75f, true);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: comment here that true is for access order.

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 2 above)
Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: No thats good. Remember the cache is still not thread safe.

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 5 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: This class is current limited for general use. For accounting memory, it now use gerSerizedSize, so it can now only works for the value of type AdHocPlannedStatement.
And it now only supported put,remove and clear methods. Other operations liked putall and iterator cannot be easily extended, (and it's not needed if used here). Should I try to extend these methods?

Comment (akhanzode) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: No lets stick to inner class.

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 2 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: MAX_LITERAL_MEM changed to use system property

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 3 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: removed

File: src/frontend/org/voltdb/ParameterSet.java (hunk 4 above)
Comment (JadenW) [inline src/frontend/org/voltdb/ParameterSet.java]: fixed

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 6 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: fixed.

File: src/frontend/org/voltdb/ParameterSet.java [hunk 7]
```diff
@@ -255,6 +255,9 @@ public static ParameterSet fromByteBuffer(ByteBuffer buffer) throws IOException
         int startPos = buffer.position();
 
         short count = buffer.getShort();
+        if (count < 0) {
+            throw new RuntimeException("Invalid parameter length " + count + " for ParameterSet." );
```
Comment (vtkstef) [inline src/frontend/org/voltdb/ParameterSet.java]: use IllegalArgumentException

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 8]
```diff
@@ -108,31 +110,18 @@ private AdHocCompilerCache() {
...
-            @Override
-            public boolean removeEldestEntry(Map.Entry<String, AdHocPlannedStatement> eldest) {
-                if (size() > MAX_LITERAL_ENTRIES) {
-                    ++m_literalEvictions;
-                    return true;
-                }
-                return false;
-            }
-        };
+        m_literalCache = new MemoryBoundedLinkedHashMap<String, AdHocPlannedStatement>(MAX_LITERAL_ENTRIES, MAX_LITERAL_MEM);
```
Comment (vtkstef) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: Don't template the MemoryBoundedLinkedHashMap call it AdHocStatementCache which extends LinkedHashMap<String, AdHocPlannedStatement>

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 9]
```diff
@@ -147,9 +136,70 @@ public boolean removeEldestEntry(Map.Entry<String, List<BoundPlan> > eldest) {
...
+                ++m_literalEvictions;
+                this.currentMemory -= ((org.voltdb.compiler.AdHocPlannedStatement) eldest.getValue()).getSerializedSize();
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public AdHocPlannedStatement put(K key, AdHocPlannedStatement value) {
+            this.currentMemory += ((org.voltdb.compiler.AdHocPlannedStatement) value).getSerializedSize();
```
Comment (vtkstef) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: why do you need to cast it?

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java [hunk 10]
```diff
@@ -147,9 +136,70 @@ public boolean removeEldestEntry(Map.Entry<String, List<BoundPlan> > eldest) {
...
+            this.maxMemory = maxMemory;
+            this.currentMemory = 0;
+        }
+
+        // This method is called just after a new entry has been added
+        @Override
+        public boolean removeEldestEntry(final Map.Entry<K, AdHocPlannedStatement> eldest) {
+            if ((size() > maxEntries) || (this.currentMemory > this.maxMemory))  {
+                ++m_literalEvictions;
+                this.currentMemory -= ((org.voltdb.compiler.AdHocPlannedStatement) eldest.getValue()).getSerializedSize();
```
Comment (vtkstef) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: why do you need to cast here?

File: src/frontend/org/voltdb/ParameterSet.java (hunk 7 above)
Comment (JadenW) [inline src/frontend/org/voltdb/ParameterSet.java]: fixed

File: src/frontend/org/voltdb/compiler/AdHocCompilerCache.java (hunk 10 above)
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/AdHocCompilerCache.java]: removed the unnecessary cast due to previous template
